### PR TITLE
Release v3.2.0

### LINKED ISSUE #553: Manta 3.2.0 Release checklist
# Release Checklist

Most of the following checks should be completed before officially publishing the new release
of the Calamari/Manta runtime or client. Some need to be completed after the new code is deployed.

### Runtime Releases

These checks should be performed on the codebase prior to freezing our release candidate:

- [x] Verify [`spec_version`](#spec-version) has been incremented since the
    last release for any native runtimes from any existing use on public
    (non-private/test) networks. If the runtime was published (release or pre-release), either
    the `spec_version` or `impl` must be bumped.
- [x] Verify pallet and [extrinsic ordering](#extrinsic-ordering) has stayed
    the same. Bump `transaction_version` if not.
- [x] Verify new extrinsics have been correctly whitelisted/blacklisted
- [x] Verify [benchmarks](#benchmarks) have been updated for any modified
    runtime logic.
- [x] Check for any upstream storage migrations and perform tests with `try-runtime`, if any.

The following checks can be performed after we have frozen our release candidate:

- [ ] Code freeze should typically happen one week prior to release, to ensure we have enough time for related testing.
- [ ] Notify everyone, especially people with merge rights to `manta` (stechu, Dengjianping) that a release is ongoing and no more merges to `manta` should happen until told otherwise
- [ ] Complete the following [manual QA workflow](https://hackmd.io/TbFmorG2RnOPmLuFcg9JOQ?view).
- [ ] Verify Polkadot JS API are up to date with the latest
    runtime changes.
- [ ] Execute runtime upgrade to Como and verify network stability.
- [ ] Execute runtime upgrade to Baikal and verify network stability.
- [ ] Prepare a governance post and submit to our forum with description and motivation for changes.

### Client Releases

- [ ] Verify that each crate's `version` has been bumped from previous release.
- [ ] Check that the new client versions have [burned-in](#burn-in) without issue for at least 12 hours.

### All Releases

- [ ] Check that a draft release has been created at
    https://github.com/Manta-Network/Manta/releases with relevant [release
    notes](#release-notes)
- [ ] Check that build artifacts have been added to the
    draft-release
- [ ] Coordinate with marketing team for documentation updates and other relevant tasks.
- [ ] Update changelog.

Note: Do not publish draft releases from PR branches, because those branches will be deleted when the PR is merged.

### After Runtime Upgrade
- [ ] Notify subscan team. Ensure subscan service can continue to scan calamari blocks.

### Burn In

Ensure that Manta DevOps has run the new release on Como and Baikal nodes
for at least 12 hours prior to publishing the release.

### Release notes

The release notes should list:

- The priority of the release (i.e., how quickly users should upgrade) - this is
    based on the max priority of any *client* changes.
- Which native runtimes and their versions are included
- The proposal hashes of the runtimes as built with
    [srtool](https://gitlab.com/chevdor/srtool)
- Any changes in this release that are still awaiting audit

The release notes may also list:

- Free text at the beginning of the notes mentioning anything important
    regarding this release
- Notable changes (those labelled with B[1-9]-* labels) separated into sections

### Spec Version

A runtime upgrade must bump the spec number. This may follow a pattern with the
client release

### Extrinsic Ordering

Offline signing libraries depend on a consistent ordering of call indices and
functions. Compare the metadata of the current and new runtimes and ensure that
the `module index, call index` tuples map to the same set of functions. To generate a diff report you can do the following:
* Go to the [metadata diff tool page](https://github.com/Manta-Network/Manta/actions/workflows/metadata_diff.yml).
* Open `Run workflow` drop-down menu.
* Choose the branch you want to test and other inputs if the defaults are not applicable.
* Run the workflow.
* Wait for the job to complete and inspect the output. The things to look for in the `output.txt` are lines like:
  - `[Identity] idx 28 -> 25 (calls 15)` - indicates the index for `Identity` has changed
  - `[+] Society, Recovery` - indicates the new version includes 2 additional modules/pallets.
  - If no indices have changed, every modules line should look something like `[Identity] idx 25 (calls 15)`

 In case of a breaking change, bump the `transaction_version`.

Note: Adding new functions to the runtime does not constitute a breaking change
as long as the indexes did not change.

### Benchmarks

There are three benchmarking machines reserved for updating the weights at
release-time. To initialize a benchmark run for each production runtime
(calamari, manta):
* Go to [Calamari Benchmarking Github Action](https://github.com/Manta-Network/Manta/actions/workflows/generate_calamari_weights_files.yml) 
  and [Manta Benchmarking Github Action](https://github.com/Manta-Network/Manta/actions/workflows/generate_manta_weights_files.yml)
* Open `Run workflow` drop-down menu.
* Choose your branch and run the workflow.
* When these jobs have completed (it takes a few hours), custom weights files will
    be available to download as artifacts. 
* Commit the changes to your branch and push to the remote branch for review.
* The weights should be (Currently manually) checked to make sure there are no
    big outliers (i.e., twice or half the weight).

### Security Audit

Before release, run a `Security Audit`

* Go to [Security Audit](https://github.com/Manta-Network/Manta/actions/workflows/audit.yml).
* Open `Run workflow` drop-down menu.
* Choose your branch and run the workflow.
* An audit report will be generated.
* Address any reported findings before release. If it cannot be fixed, please file an issue.

### PR DESCRIPTION
**Pull request by Dengjianping**
Signed-off-by: Dengjianping <djptux@gmail.com>

## Description

closes: #553

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests.
- [x] Updated relevant documentation in the code.
- [x] Added **one** line describing your change in `<branch>/CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [x] If runtime changes, need to update the version numbers properly:
   * `authoring_version`: The version of the authorship interface. An authoring node will not attempt to author blocks unless this is equal to its native runtime.
   * `spec_version`: The version of the runtime specification. A full node will not attempt to use its native runtime in substitute for the on-chain Wasm runtime unless all of spec_name, spec_version, and authoring_version are the same between Wasm and native.
   * `impl_version`: The version of the implementation of the specification. Nodes are free to ignore this; it serves only as an indication that the code is different; as long as the other two versions are the same then while the actual code may be different, it is nonetheless required to do the same thing. Non-consensus-breaking optimizations are about the only changes that could be made which would result in only the impl_version changing.
   * `transaction_version`: The version of the extrinsics interface. This number must be updated in the following circumstances: extrinsic parameters (number, order, or types) have been changed; extrinsics or pallets have been removed; or the pallet order in the construct_runtime! macro or extrinsic order in a pallet has been changed. You can run the `metadata_diff.yml` workflow for help. If this number is updated, then the `spec_version` must also be updated 
- [x] Verify benchmarks & weights have been updated for any modified runtime logics
- [x] If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- [x] If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- [x] If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.

### CODE REVIEW

**Comment by Dengjianping** [comment]
Wait #554 to be merged.

**Comment by ghzlatarev** [comment]
Are you going to modify the Sudo migration to drain the `pallet-version` storage ?

**Comment by Dengjianping** [comment]
> Are you going to modify the Sudo migration to drain the `pallet-version` storage ?

Sure, I'm testing it locally now.

**Review by stechu — DISMISSED** [review]

**Comment by stechu** [comment]
LGTM. need to merge #529 first tho.

**Review by ghzlatarev — DISMISSED** [review]

**Comment by Garandor** [comment]
holding off on review until #529 is merged

**Review by Garandor — CHANGES_REQUESTED** [review]
Asset-manager needs a `try-runtime` implementation in pallets/asset-manager/src/lib.rs:567

**Comment by ghzlatarev** [comment]
@Dengjianping there's one more issue with https://github.com/Manta-Network/Manta/blob/manta/pallets/asset-manager/src/lib.rs#L535  It should be
```
Junctions::X2(Junction::Parachain(para_id), _) => {
					AllowedDestParaIds::<T>::contains_key(para_id)
				}
```
Otherwise we won't be able to send to AccountKey20 types for example

**Comment by stechu** [comment]
> @Dengjianping there's one more issue with https://github.com/Manta-Network/Manta/blob/manta/pallets/asset-manager/src/lib.rs#L535 It should be
> 
> ```
> Junctions::X2(Junction::Parachain(para_id), _) => {
> 					AllowedDestParaIds::<T>::contains_key(para_id)
> 				}
> ```
> 
> Otherwise we won't be able to send to AccountKey20 types for example

Good catch

**Comment by Dengjianping** [comment]
> Asset-manager needs a `try-runtime` implementation in pallets/asset-manager/src/lib.rs:567

Not sure what you mean.

There's `on_runtime_upgrade` hook in asset-manager, `try-runtime` can find it automatically, imo.
This is the log from try-runtime CI:
```
AssetManager declares internal migrations (which *might* execute). On-chain `StorageVersion(0)` vs current storage version `StorageVersion(0)`  
```

Perhaps I have to bump pallet-version?

**Comment by Garandor** [comment]
> > Asset-manager needs a `try-runtime` implementation in pallets/asset-manager/src/lib.rs:567
> 
> Not sure what you mean.
> 
> There's `on_runtime_upgrade` hook in asset-manager

the actual runtime upgrade must not fail/panic, so if the upgrade fails to apply cleanly, `on_runtime_upgrade` will just ignore the error.

We want the migration to do two additional things:

- [ ]  have a pre/postcondition check that can simulate the upgrade and panic (or at least print failure) if the function in `on_runtime_upgrade` would throw an error. This can be done with pre/post migration steps used with `try_runtime`, as in pallets/collator-selection/src/migrations.rs:68
- [ ] Ensure the migration only executes once, unless you can guarantee that it's invariant on reexecution. Easiest thing to do is check against `pallet_version` and bump it after the upgrade, or check the specific runtime version

see also
```
#[cfg(feature = "try-runtime")]
fn pre_upgrade() -> Result<(), &'static str> { Ok(()) }

#[cfg(feature = "try-runtime")]
fn post_upgrade() -> Result<(), &'static str> { Ok(()) }
```
in https://docs.substrate.io/v3/tools/try-runtime/

File: pallets/asset-manager/src/lib.rs
Comment (ghzlatarev) [inline pallets/asset-manager/src/lib.rs:540]: We'll probably end up needing to support more junctions than `AccountKey20` like `AccountIndex64` . That's why I suggested:
```
Junctions::X2(Junction::Parachain(para_id), _) => { 
     AllowedDestParaIds::<T>::contains_key(para_id)
}
``` 
But I guess let's cross that bridge when we get there.

**Review by ghzlatarev — DISMISSED** [review]

File: pallets/asset-manager/src/lib.rs
Comment (Garandor) [inline pallets/asset-manager/src/lib.rs:561]: Is LocationAssetId continuing to be in use after the upgrade of not?

If not, please drain it and check it exists in the pre- and doesn't exist anymore in the post check

**Comment by Garandor** [comment]
I'll schedule #460 for the 3.2.1 release to clean up the old PalletVersion storage that is becoming unused as we migrate to storageVersion

see https://paritytech.github.io/substrate/master/src/frame_support/migrations.rs.html#36 of https://paritytech.github.io/substrate/master/frame_support/migrations/fn.migrate_from_pallet_version_to_storage_version.html

File: CHANGELOG.md
Comment (Garandor) [inline CHANGELOG.md]: is this a breaking change? What exactly does it break for a user?

**Review by Garandor — COMMENTED** [review]
LGTM besides my comments

File: pallets/asset-manager/src/lib.rs
Comment (Dengjianping) [reply pallets/asset-manager/src/lib.rs:561]: `LocationAssetId` is in use.

File: CHANGELOG.md
Comment (Dengjianping) [reply CHANGELOG.md]: It's important, because we're going to remove sudo from calamari.
Let me move it to `Improvements`.

Comment (Garandor) [reply CHANGELOG.md]: Sounds good. The "Breaking" section is really not about importance of the change, but whether someone who uses the node (e.g. frontend, polkadot.js) has to change their code or some functionality they used to have will no longer work after this node upgrade, so stuff like metadata changes or changes of extrinsics to have different parameters.

I think you could argue that removing the "execute with sudo" extrinsic can break some script that depends on it, but since sudo is not user-facing, I don't think it qualifies

**Comment by stechu** [comment]
pending #583 's merge.

File: pallets/asset-manager/src/lib.rs
Comment (Dengjianping) [reply pallets/asset-manager/src/lib.rs:540]: For now, I didn't see any use case on `AccountIndex64`.

**Review by ghzlatarev — DISMISSED** [review]

**Review by Garandor — DISMISSED** [review]
LGTM, thanks!

**Review by Garandor — APPROVED** [review]

**Review by ghzlatarev — APPROVED** [review]

**Pull request merged**
